test/functional/normal/put_spec.lua

```diff
@@ -0,0 +1,1012 @@
+        -- the buffer instead of using them to move after the insert has been
+        -- done.
+        -- I could probably get this working properly with a special flag put
+        -- into start_redo_ins() and set in do_put(), but I think this adds
+        -- much more complexity than fixing this bug justifies.
+        pending('should not change the ". register with ".2p', function()
+            local orig_register = funcs.getreg('.')
+            feed('2".p')
+            eq(orig_register, funcs.getreg('.'))
+        end)
+        describe("cursor positioning after undo and redo with '.'", function()
+            before_each(reset)
+            local function make_cursor_test(macro_string)
+                return function()
+                    feed(macro_string)
+                    local afterpos = funcs.getcurpos()
+                    local orig_string = curbuf_contents()
+                    feed('u.')
+                    eq(afterpos, funcs.getcurpos())
+                    expect(orig_string)
+                end
+            end
+            -- The difficulty here is: setting the cursor after the end of the
+            -- pasted text is done by adding a motion command to the
+            -- stuffbuffer after the insert.
+            -- Modifying 'redobuff' is done in the code that handles inserting
+            -- text and moving around.
+            -- I could add a special case in ins_esc() that checks for a flag
+            -- set in do_put() to add the motion character to the redo buffer,
+            -- but I think that is starting to get way to convoluted for the
+            -- benefit.
+            pending('should be the same after ".gp and ".gpu.',
+                    make_cursor_test('".gp'))
+            -- The difficulty here is: putting forwards is implemented by using
+            -- 'a' instead of 'i' to start insert.
+            -- Undoing with 'u' an insert that began with 'a' leaves the cursor
+            -- where the first character was inserted, not where the cursor was
+            -- when the 'a' was pressed.
+            -- We account for this the first time by saving the cursor position
+            -- in do_put(), but this isn't stored in redobuff for a second time
+            -- around.
+            -- We can't change how such a fundamental action as undo after
+            -- inserting with 'a' behaves, we could add in a special case
+            -- whereby we set a flag in do_put() and read it when entering
+            -- insert mode but this seems like way too much to fix such a minor
+            -- bug.
+            pending('should be the same after ".pu. and ".pu.u.',
+                    make_cursor_test('".pu.'))
+        end)
+        -- The difficulty here is: The method used to put the dot register into
+        -- the buffer is by putting keys into the readbuff, in order to avoid
+        -- autoindent I would have to either find a vimL way of avoiding it and
```

Comment:
An ugly, but perhaps isolated, way to do it in viml might be `:set paste ..... :set nopaste`. (Let's keep that in a separate commit if you try it)